File: projects/g-api-auth-httplib2/fuzz_http.py
#!/usr/bin/python3
# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import sys
import atheris

import six
from six.moves import http_client
import mock
import httplib2
import google_auth_httplib2


# Mocks for fuzzing. Inspired by testing infra. Main difference is we have
# added logic to handle the event where the MockHTTP has run out of
# responses an appropriate exception is thrown that we catch in the fuzzer.
class FuzzMockHttp(object):
    def __init__(self, responses, headers=None):
        self.responses = responses
        self.requests = []
        self.headers = headers or {}
        self.add_certificate = mock.Mock(return_value=None)

    def request(
        self,
        url,
        method="GET",
        body=None,
        headers=None,
        redirections=httplib2.DEFAULT_MAX_REDIRECTS,
        connection_type=None,
    ):
        self.requests.append(
            (method, url, body, headers, redirections, connection_type)
        )
        if len(self.responses) == 0:
            raise Exception("FUZZ: No more responses")
        return self.responses.pop(0)


class FuzzMockResponse(object):
    def __init__(self, status=http_client.OK, data=b""):
        self.status = status
        self.data = data

    def __iter__(self):
        yield self
        yield self.data


class FuzzMockCredentials(object):
    def __init__(self, token="token"):
        self.token = token

    def apply(self, headers):
        headers["authorization"] = self.token

    def before_request(self, request, method, url, headers):
        self.apply(headers)

    def refresh(self, request):
        self.token += "1"


def TestOneInput(data):
    fdp = atheris.FuzzedDataProvider(data)

    mock_credentials = mock.Mock(wraps=FuzzMockCredentials(fdp.ConsumeUnicodeNoSurrogates(10)))
    if fdp.ConsumeBool() == True:
        mock_credentials.apply({'authorization': ""})

    # Create responses. We need at least two.
    http_codes = [400, 401, 403, 404, 405, 413, 415, 429, 500, 503, 504]
    mock_responses = []
    for i in range(fdp.ConsumeIntInRange(1, 10)):
        mock_responses.append(
            FuzzMockResponse(
                status = http_codes[fdp.ConsumeIntInRange(0, len(http_codes)-1)],
                data = fdp.ConsumeBytes(100)
            )
        )
    mock_http = FuzzMockHttp(mock_responses)

    authed_http = google_auth_httplib2.AuthorizedHttp(
        mock_credentials, http=mock_http
    )
    try:
        r, d = authed_http.request(
            "http://localhost:8001",
            fdp.ConsumeUnicodeNoSurrogates(10),
            None,
            None,
            httplib2.DEFAULT_MAX_REDIRECTS,
            None
        )
    except Exception as e:
        if "FUZZ" in str(e):
            pass
        else:
            raise e


def main():
    atheris.instrument_all()
    atheris.Setup(sys.argv, TestOneInput, enable_python_coverage=True)
    atheris.Fuzz()


if __name__ == "__main__":
    main()

File: projects/unit/fuzzer/Fuzz_http.c
/* Copyright 2022 Google LLC
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "Fuzz_http.h"

#define kMinInputLength 10
#define kMaxInputLength 5120
static int DoInit = 0;

nxt_lvlhsh_t hash;

extern char  **environ;

extern int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size)
{//src/test/nxt_http_parse_test.c

    if (Size < kMinInputLength || Size > kMaxInputLength){
        return 0;
    }

    if(!DoInit){
        nxt_lib_start("tests", NULL, &environ);
        nxt_memzero(&hash, sizeof(nxt_lvlhsh_t));

        nxt_http_fields_hash(&hash, nxt_h1p_fields,
                                nxt_nitems(nxt_h1p_fields));
        DoInit = 1;
    }

    nxt_str_t nxt_http_request;

    nxt_http_request.length = Size;
    nxt_http_request.start = (uint8_t *) Data;

    return nxt_http_parse_fuzz(&nxt_http_request, &hash);
}

nxt_int_t nxt_http_parse_fuzz(nxt_str_t *request, nxt_lvlhsh_t *hash){

    nxt_mp_t                  *mp;
    nxt_buf_mem_t             buf;
    nxt_http_request_parse_t  rp;

    buf.start = request->start;
    buf.end = request->start + request->length;

    nxt_memzero(&rp, sizeof(nxt_http_request_parse_t));

    mp = nxt_mp_create(1024, 128, 256, 32);

    nxt_http_parse_request_init(&rp, mp);

    buf.pos = buf.start;
    buf.free = buf.end;

    if (nxt_slow_path(nxt_http_parse_request(&rp, &buf) == NXT_DONE)) {
        nxt_http_fields_process(rp.fields, hash, NULL);
    }

    nxt_mp_destroy(mp);

    return NXT_OK;
}
